src/coreComponents/constitutive/solid/ModifiedCamClay.hpp

```diff
@@ -0,0 +1,568 @@
+                                   localIndex const q ) const override final
+  {
+    ElasticIsotropicPressureDependentUpdates::saveConvergedState( k, q );
+    m_oldPreConsolidationPressure[k][q] = m_newPreConsolidationPressure[k][q];
+  }
+
+private:
+
+  /// A reference to the ArrayView holding the virgin compression index for each element.
+  arrayView1d< real64 const > const m_virginCompressionIndex;
+
+  /// A reference to the ArrayView holding the slope of the critical state line for each element.
+  arrayView1d< real64 const > const m_cslSlope;
+
+  /// A reference to the ArrayView holding the new preconsolidation pressure for each integration point
+  arrayView2d< real64 > const m_newPreConsolidationPressure;
+
+  /// A reference to the ArrayView holding the old preconsolidation presure for each integration point
+  arrayView2d< real64 > const m_oldPreConsolidationPressure;
+
+};
+
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void ModifiedCamClayUpdates::evaluateYield( real64 const p,
+                                            real64 const q,
+                                            real64 const pc,
+                                            real64 const M,
+                                            real64 const Cc,
+                                            real64 const Cr,
+                                            real64 const bulkModulus,
+                                            real64 const mu,
+                                            real64 & f,
+                                            real64 & df_dp,
+                                            real64 & df_dq,
+                                            real64 & df_dpc,
+                                            real64 & df_dp_dve,
+                                            real64 & df_dq_dse ) const
+{
+
+  df_dp = -pc + 2. * p;
+  df_dq = 2. * q /(M*M);
+  df_dpc = -p;
+  real64 dpc_dve = -1./(Cc-Cr) * pc;
+  df_dp_dve = 2. * bulkModulus + dpc_dve;
+  df_dq_dse = 2. /(M*M) * 3. * mu;
+
+  f = q*q/(M*M)+p*(p-pc);
+
+}
+
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void ModifiedCamClayUpdates::smallStrainUpdate( localIndex const k,
+                                                localIndex const q,
+                                                real64 const ( &strainIncrement )[6],
+                                                real64 ( & stress )[6],
+                                                real64 ( & stiffness )[6][6] ) const
+{
+
+  // Rename variables for easier implementation
+
+  real64 const oldPc  = m_oldPreConsolidationPressure[k][q];   //pre-consolidation pressure
+  real64 const mu     = m_shearModulus[k];
+  real64 const p0     = m_refPressure;
+
+  real64 const eps_v0 = m_refStrainVol;
+  real64 const M      = m_cslSlope[k];
+  real64 const Cr     = m_recompressionIndex[k];
+  real64 const Cc     = m_virginCompressionIndex[k];
+
+  real64 pc    = oldPc;
+  real64 bulkModulus  = -p0/Cr;
```

Comment:
We may want to define material models that relate elastic moduli to pressure, porosity, damage, etc.. Such model can be called by any of the plastic model (MCC, Delft-Egg, DP, etc.)

Reply:
Totally agree with you and we already have similar discussion after the validation tests of the implemented Delft-Egg/MCC model. I would suggest that we finalize this PR with classic models and then think about further development with a more generalized elastic module (expand the hyper-elastic one) as a new PR.

Reply:
This is a nice idea, but we have to balance generality with performance.